File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        result = 0
        max_freq = 0
        freq = {}

        left = 0
        for right in range(0, len(s)):
            char = s[right]
            if char not in freq:
                freq[char] = 1
            else:
                freq[char] += 1

            max_freq = max(max_freq, freq[char])

            while (right - left + 1) - max_freq > k:
                freq[s[left]] -= 1
                left += 1
            
            result = max(result, right - left + 1)

        return result
        

File: Data Structures & Algorithms/longest-substring-with-at-most-two-distinct-characters/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        int longest_substr_length = INT_MIN;
        unordered_map<char, int> freq_map;

        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            char right_char = s[right];

            freq_map[right_char]++;

            while (freq_map.size() > 2) {
                char left_char = s[left];
                freq_map[left_char]--;

                if (freq_map[left_char] == 0) {
                    freq_map.erase(left_char);
                }
                
                left += 1;
            }

            int curr_substr_length = right - left + 1;
            longest_substr_length = max(longest_substr_length, curr_substr_length);

        }
        return longest_substr_length;
    }
};